### ISSUE TITLE
Incompatible with KeePass 2.40

### ISSUE THREAD

**Issue by stbxxl** [Kyrodan/KeeAnywhere#141]
Hi,

It appears KeeAnywhere 1.41 is not compatible with the latest KeePass version 2.40. When I start KeePass 2.40 I get the attached error message for the KeeAnywhere plugin.

I hope it's possible to make this very useful plugin work again.

![keeanywhere error keepass 2 40](https://user-images.githubusercontent.com/19540455/45305180-9c19cf80-b4ce-11e8-86e0-5dc9a44966e5.jpg)

**Comment by chakal** [comment]
Same here, appears to be fixed in the dev repo according to fixed dependencies, but I don't have a compiling environement to test

**Comment by gjf** [comment]
I can confirm - same issue

**Comment by christantoan** [comment]
Workaround that I found (I don't know if it's safe though):
1. Download [this file](https://www.nuget.org/api/v2/package/Microsoft.Graph.Core/1.3.1)
2. Rename its extension to .zip
3. Extract Microsoft.Graph.Core.dll from microsoft.graph.core.1.3.1.nupkg.zip\lib\portable45-net45+win8+wpa81 folder to where your KeePass.exe is located

**Comment by Yastre** [comment]
I have the same problem, i had change to other plugin for keep updated my keepass whit the cloud storage

**Comment by DaftVader43** [comment]
My Workaround
Removed the plugin, restarted Keepass. 
Closed Keypass, added plugin back in, opened Keepass... worked fine.

I stumbled across this - I removed it thinking "oh well :/" , tried some other plugins (no luck..and added it back to try again.

**Comment by stefan-lindvall** [comment]
@DaftVader43: Your workaround did not work for me.
@christantoan: This workaround worked great for me.

**Comment by schillich** [comment]
Thanks @christantoan for this workaround, it works like a charm.

**Comment by CommanderBond** [comment]
Thanks @christantoan. Works for me as well.

**Comment by CrendKing** [comment]
I think all you need is the manually clear the plugin cache of the plugin. When you open KeePass, it will regenerate the cache files, which include the compatible version of Microsoft.Graph.Core.dll.

**Comment by Ennneerrrgggyyy** [comment]
@CrendKing that didn't work for me 🙁

**Comment by gjf** [comment]
It looks like workaround is as follows.
1. Close KeePass.
2. Remove Keeanywhere folder from the cache.
3. Place the file @christantoan mentioned in KeePass folder (not plugin or cache!)
4. Try to start KeePass again. It will work with new Microsoft.Graph.Core.dll and the plugin installs well.

**Comment by Kyrodan** [comment]
Could you please try this short (unstable) developer build and give me feedback?
But please undo any "hacks" on your KeePass installation first (like deleting any copied DLLs in KeePass' Program Folder).

The attached file is a zipped PLGX - so please unzip at first and copy the PLGX to KePass' plugins folder.

[KeeAnywhere-1.5.0-unstable-1.plgx.zip](https://github.com/Kyrodan/KeeAnywhere/files/2371162/KeeAnywhere-1.5.0-unstable-1.plgx.zip)

**Comment by bryanmcain** [comment]
The developer build works for me, thank you!

**Comment by schillich** [comment]
The developer build works for me too, thanks.

**Comment by cathcam** [comment]
is the plugins folder %LocalAppData%\KeePass\PluginCache

**Comment by Kyrodan** [comment]
@cathcam: no, ist %programfiles(x86)%\KeePass Password Safe 2\Plugins

**Comment by cathcam** [comment]
Hmm, not sure how it got set, but I needed to put the plugin in %programfiles(x86)%\KeePass Password Safe 2

and rename the KeeAnywhere-1.4.1.plgx to .old to get it to work, but it worked first time after that, I got the upgraded window and it loaded from dropbox. Thanks!

**Comment by dragoangel** [comment]
@Kyrodan dev build work with DropBox, tested.

**Comment by ReechK** [comment]
I can confirm KeeAnywhere-1.5.0-unstable-1 works with 2.40 + dropbox://

**Comment by bluegeekgh** [comment]
The developer build works for me with Google Drive.

**Comment by mcolegro** [comment]
The 1.5-unstable version works for me with DropBox.

**Comment by Nukezz** [comment]
Can also confirm 1.5-unstable worked (Google Drive). Had to remove current release plugin, start KeePass, close, add 1.5-unstable, then start KeePass again to kick it over.

**Comment by vbarnar** [comment]
Works with Onedrive!

**Comment by stbxxl** [comment]
I can confirm that the 1.5-unstable build works with Dropbox. Local backups also work.

**Comment by n0-bs** [comment]
1.5-unstable build works with AWS S3

**Comment by StormykeithDMFG** [comment]
Working well here with Google Drive

**Comment by auanasgheps** [comment]
1.5-unstable build works fine on W10, Keepass 2.40, Google Drive. Thanks!

**Comment by mcfisch** [comment]
v1.5.0 does not work for me on a domain joined Windows 10 Enterprise. However, the error itself has changed after replacing the plugin:
```
The following plugin is incompatible with the current KeePass version:
C:\Users\mcfisch\AppData\Local\KeePass\PluginCache\Ki9k42VfgyEUGpS39j3w\KeeAnywhere.dll

Have a look at the plugin's website for an appropriate version.

[Expanded Information]
Key not valid for use in specified state.

System.Security
   at System.Security.Cryptography.ProtectedData.Unprotect(Byte[] encryptedData, Byte[] optionalEntropy, DataProtectionScope scope)
   at KeeAnywhere.Json.ProtectedDataStringPropertyResolver.ProtectedDataStringValueProvider.SetValue(Object target, Object value)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateList(IList list, JsonReader reader, JsonArrayContract contract, JsonProperty containerProperty, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateList(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, Object existingValue, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at KeeAnywhere.Configuration.ConfigurationService.LoadAccountsFromLocalSecureStore()
   at KeeAnywhere.Configuration.ConfigurationService.Load()
   at KeeAnywhere.KeeAnywhereExt.Initialize(IPluginHost pluginHost)
   at KeePass.Plugins.PluginManager.LoadPlugins(List`1 lFiles, String strTypeName, String strDisplayFilePath, Boolean bSkipCacheFiles)
Byte[] Unprotect(Byte[], Byte[], System.Security.Cryptography.DataProtectionScope)
```
Unfortunately this computer was just migrated to another active directory domain, which caused a whole lot of other issues, so I'm not fully convinced this isn't related as well.

**Comment by cbeganskas** [comment]
KeeAnywhere-1.5.0-unstable-1.plgx restored functionality under Google Drive.

**Comment by Biro0702** [comment]
The KeeAnywhere-1.5.0-unstable-1.plgx works fine with dropbox for me. Thanks!!
Steps used: 
* Close Keepass.
* del KeeAnywhere-1.4.1.plgx from plugin folder.
* put KeeAnywhere-1.5.0-unstable-1.plgx into plugin folder.
* Open Keepass.
Everything works fine.

**Comment by erreip** [comment]
> Could you please try this short (unstable) developer build and give me feedback?
> But please undo any "hacks" on your KeePass installation first (like deleting any copied DLLs in KeePass' Program Folder).
> 
> The attached file is a zipped PLGX - so please unzip at first and copy the PLGX to KeePass' plugins folder.
> 
> [KeeAnywhere-1.5.0-unstable-1.plgx.zip](https://github.com/Kyrodan/KeeAnywhere/files/2371162/KeeAnywhere-1.5.0-unstable-1.plgx.zip)

Workaround Worked for me too. thanks @Kyrodan

**Comment by big-bad-wolfe** [comment]
unstable also works for me.

**Comment by Malevola** [comment]
Unstable works with Dropbox.

**Comment by vaindil** [comment]
Unstable does not work for me on a completely fresh Windows/KeePass install. I don't know where the log file is to copy/paste the stack trace, so here's a screenshot. The main message is "Key not valid for use in specified state."

![stacktrace](https://user-images.githubusercontent.com/771134/45596743-da136980-b98e-11e8-8954-8c38c5d02e27.png)

**Comment by Stanislasj** [comment]
Sorry but not working for me with unstable + KP2.40 under WIN 7. I del all plugin, relaunch, and this message: No link to dropbox
![image](https://user-images.githubusercontent.com/43320576/45598957-e1b12d80-b9e3-11e8-8acd-499af6bc93d8.png)

**Comment by Gedankenleser** [comment]
KeeAnywhere-1.5.0-unstable-1.plgx works fine!

**Comment by lgs4gh** [comment]
Unstable 1.5.0 works for me, google drive, Win 10 Pro.

**Comment by gonik** [comment]
I can confirm that Unstable 1.5.0 works for OneDrive as well. For some reason, just replacing the plgx did not work. I had to remove all files created (settings, cache, saved passwords) by the plugin and re-authorize OneDrive in order for it to work.

**Comment by Hatshepam** [comment]
> Workaround that I found (I don't know if it's safe though):
> 
> 1. Download [this file](https://www.nuget.org/api/v2/package/Microsoft.Graph.Core/1.3.1)
> 2. Rename its extension to .zip
> 3. Extract Microsoft.Graph.Core.dll from microsoft.graph.core.1.3.1.nupkg.zip\lib\portable45-net45+win8+wpa81 folder to where your KeePass.exe is located
> 
> Your final result should be something like this:
> ![image](https://user-images.githubusercontent.com/7300310/45314036-f5b2e600-b530-11e8-9f27-341c810254ac.png)
> 
> PS:
> @chakal, I tested [the latest version from the dev repo](https://ci.appveyor.com/project/Kyrodan/keeanywhere/build/1.0.0.121-develop/artifacts). It still has other problems with actually getting the files from my Google Drive.

So far your solution has worked perfectly for me, thank you!

**Comment by cgabba** [comment]
KeeAnywhere-1.5.0-unstable-1.plgx works for me. Tnx

**Comment by cgabba** [comment]
KeeAnywhere-1.5.0-unstable-1.plgx works for me. Tnx

**Comment by oliverclozov** [comment]
1.50 unstable plugin confirmed working on Win 10

**Comment by reillychase** [comment]
@Kyrodan KeeAnywhere-1.5.0-unstable-1.plgx.zip is working for me, thanks

**Comment by iesoftwaredeveloper** [comment]
The problem can be fixed similar to previous errors with updates to KeePass.  Open the KeePass.exe.config file.  Find the following code:

`<assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
			<dependentAssembly>
				<assemblyIdentity name="KeePass"
					publicKeyToken="fed2ed7716aecf5c"
					culture="neutral" />
				<bindingRedirect oldVersion="2.0.9.0-2.40.0.0"
					newVersion="2.40.0.17978" />
			</dependentAssembly>
      <dependentAssembly>
		</assemblyBinding>
`

And add to it  bindingRedirect for the Microsoft.Graph.Core.  After the add, it should look like this.
`		<assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
			<dependentAssembly>
				<assemblyIdentity name="KeePass"
					publicKeyToken="fed2ed7716aecf5c"
					culture="neutral" />
				<bindingRedirect oldVersion="2.0.9.0-2.40.0.0"
					newVersion="2.40.0.17978" />
			</dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="Microsoft.Graph.Core" publicKeyToken="31bf3856ad364e35" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-1.4.0.0" newVersion="1.4.0.0" />
      </dependentAssembly>
		</assemblyBinding>
`

**Comment by Kyrodan** [comment]
Released in v1.5.0

**Comment by marcelser** [comment]
Tried solution from iesoftwaredeveloper of changing KeePass.exe.config with above change. I still get the error on plugin change
